File: Phan4/bai4.3/bai4.3.jl
board = [
    0 0 0 0 0 6 0 0 0;
    0 0 8 1 0 5 0 4 0;
    0 0 0 9 0 0 7 0 5;
    0 0 0 0 0 9 0 0 0;
    0 0 0 6 0 0 1 0 0;
    0 7 4 0 0 0 0 2 1;
    0 0 0 0 1 3 0 0 0;
    0 0 0 0 0 2 9 0 0;
    0 9 0 0 0 0 8 0 6
]

function is_valid(b, r, c, val)
    for i in 1:9
        if b[r, i] == val || b[i, c] == val
            return false
        end
    end
    start_r, start_c = 3 * div(r - 1, 3) + 1, 3 * div(c - 1, 3) + 1
    for i in 0:2, j in 0:2
        if b[start_r + i, start_c + j] == val
            return false
        end
    end
    return true
end

global count_seq = 0
function solve_seq!(b)
    global count_seq += 1
    for r in 1:9, c in 1:9
        if b[r, c] == 0
            for val in 1:9
                if is_valid(b, r, c, val)
                    b[r, c] = val
                    if solve_seq!(b) return true end
                    b[r, c] = 0
                end
            end
            return false
        end
    end
    return true
end

global count_mrv = 0
function solve_mrv!(b)
    global count_mrv += 1
    min_cand = 10
    best_r, best_c = 0, 0
    
    for r in 1:9, c in 1:9
        if b[r, c] == 0
            cand = sum(is_valid(b, r, c, v) for v in 1:9)
            if cand < min_cand
                min_cand = cand
                best_r, best_c = r, c
            end
        end
    end
    
    if best_r == 0 return true end
    
    for val in 1:9
        if is_valid(b, best_r, best_c, val)
            b[best_r, best_c] = val
            if solve_mrv!(b) return true end
            b[best_r, best_c] = 0
        end
    end
    return false
end

b1 = copy(board)
t_start = time()
solve_seq!(b1)
time_seq = time() - t_start

b2 = copy(board)
t_start = time()
solve_mrv!(b2)
time_mrv = time() - t_start

is_same = (b1 == b2)

println("=== KẾT QUẢ BÀI 4.3 (JULIA) ===")
println("Số lời gọi (Quét tuần tự): $count_seq | (MRV): $count_mrv")
println("Thời gian (Quét tuần tự): $(round(time_seq, digits=4))s | (MRV): $(round(time_mrv, digits=4))s")
println("Lời giải tìm được có giống nhau không? $(is_same ? "Có" : "Không")")

File: Phan4/bai4.3/bai4.3.py
import time
import copy

# Đề Sudoku mục E (56 ô trống được điền số 0)
board = [
    [0, 0, 0, 0, 0, 6, 0, 0, 0],
    [0, 0, 8, 1, 0, 5, 0, 4, 0],
    [0, 0, 0, 9, 0, 0, 7, 0, 5],
    [0, 0, 0, 0, 0, 9, 0, 0, 0],
    [0, 0, 0, 6, 0, 0, 1, 0, 0],
    [0, 7, 4, 0, 0, 0, 0, 2, 1],
    [0, 0, 0, 0, 1, 3, 0, 0, 0],
    [0, 0, 0, 0, 0, 2, 9, 0, 0],
    [0, 9, 0, 0, 0, 0, 8, 0, 6]
]

def is_valid(b, r, c, val):
    for i in range(9):
        if b[r][i] == val or b[i][c] == val:
            return False
    start_r, start_c = 3 * (r // 3), 3 * (c // 3)
    for i in range(3):
        for j in range(3):
            if b[start_r + i][start_c + j] == val:
                return False
    return True

# 1. Quét tuần tự
count_seq = 0
def solve_sequential(b):
    global count_seq
    count_seq += 1
    for r in range(9):
        for c in range(9):
            if b[r][c] == 0:
                for val in range(1, 10):
                    if is_valid(b, r, c, val):
                        b[r][c] = val
                        if solve_sequential(b):
                            return True
                        b[r][c] = 0
                return False
    return True

# 2. Heuristic MRV
count_mrv = 0
def solve_mrv(b):
    global count_mrv
    count_mrv += 1
    
    min_cand = 10
    best_r, best_c = -1, -1
    for r in range(9):
        for c in range(9):
            if b[r][c] == 0:
                cand = sum(1 for val in range(1, 10) if is_valid(b, r, c, val))
                if cand < min_cand:
                    min_cand = cand
                    best_r, best_c = r, c
                    
    if best_r == -1:
        return True
        
    for val in range(1, 10):
        if is_valid(b, best_r, best_c, val):
            b[best_r][best_c] = val
            if solve_mrv(b):
                return True
            b[best_r][best_c] = 0
    return False

# Thực thi và đo đạc
b1 = copy.deepcopy(board)
t0 = time.time()
solve_sequential(b1)
t1 = time.time()
time_seq = t1 - t0

b2 = copy.deepcopy(board)
t0 = time.time()
solve_mrv(b2)
t1 = time.time()
time_mrv = t1 - t0

is_same = (b1 == b2)

print("=== KẾT QUẢ BÀI 4.3 (PYTHON) ===")
print(f"Số lời gọi (Quét tuần tự): {count_seq} | (MRV): {count_mrv}")
print(f"Thời gian (Quét tuần tự): {time_seq:.4f}s | (MRV): {time_mrv:.4f}s")
print(f"Lời giải tìm được có giống nhau không? {'Có' if is_same else 'Không'}")